### PR TITLE
#3: Convert documentation to GitBook

### PR DESCRIPTION
**Pull request by jonas**
Example here: https://jonas.github.io/eigenflow/

The next step will be to figure out how to automatically generate documentation using Travis and push it to the ypg-data.github.io repo to avoid having an `gh-pages` branch in this repo.

### CODE REVIEW

File: devops/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# DevOps
+
+For a quick start on Mac OS use the `devops/mac/eigenflow` script:
+
+Check docker installation and setup ports forwarding
```
Comment (yawaramin) [inline devops/README.md]: s/ports/port/

File: devops/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+```
+
+Start containers
+```
+$ ./eigenflow start
+```
+Note: The state will be reset on every start (empty database and message queues).
+
+
+Check docker container the status
```
Comment (yawaramin) [inline devops/README.md]: s/the//

File: doc/error-handling.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+# Error Handling
+
+By default, if no strategy is defined, when an exception happens the process switches to the `failed` stage remembering
+the stage it was trying to execute and exists.
```
Comment (yawaramin) [inline doc/error-handling.md]: s/ and exists/, and exits/

File: doc/error-handling.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+The simplest possible strategy would be: no matter what happens retry n-times with given interval:
```
Comment (yawaramin) [inline doc/error-handling.md]: `s/n-/_n_ /`

File: doc/getting-started.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+all stages must extend `ProcessStage` trait.
+
+Create a process class:
+
+```scala
+class SimpleProcess extends StagedProcess {
+}
+```
+
+The `StagedProcess` will require to define `executionPlan` method, this method describes the pipeline stages flow.
```
Comment (yawaramin) [inline doc/getting-started.md]: s/to define/defining/

File: doc/process-stages.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+Here is an example of downloading and archiving a report.
+
+```scala
+case object Initial extends ProcessStage
+case object Download extends ProcessStage
+case object Archive extends ProcessStage
+
+val init = Initial withContext { context: ProcessContext =>
+  Future {
+    context.processingDate.minusDays(1) // means always download report from yesterday (relating to the current processingDate)
```
Comment (yawaramin) [inline doc/process-stages.md:40]: Hmm, according to #7 `processingDate` is not yet a `LocalDateTime`, so this particular line is not yet valid. Perhaps replace with `def dayBefore(date: Date) = ...; dayBefore(context.processingDate)`?

Also we'd need to add a comment on #7 reminding us to change this back to `context.processingDate minusDays 1`.

File: doc/time-management.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+* Catching up, for example if a process should run daily but the last successful run was 3 days ago it will automatically
+run for every missing day until now.
+* Protects from double run, if a process should run daily and it **successfully** finished execution today it won't run for today
+again even if it was executed again, unless it was explicitly asked to do so.
+Thus you shouldn't worry about occasional loading of the same data twice.
+Note: to have a guaranteed double run protection, the environment must satisfy 2 conditions:
+    1. Persistence layer must not be eventual consistent. We use cassandra configured for consistency.
+    1. No simultaneous runs of the same process. We use mesos with chronos to schedule processes in cluster.
+
+Override `nextProcessingDate(lastCompleted: Date): Date` method to define next run date should be, based on the last completed date.
```
Comment (yawaramin) [inline doc/time-management.md:14]: This will also need to be updated when #7 is finished.

File: README.md [hunk 8]
```diff
@@ -26,10 +28,10 @@ val analyze = Analyze { parquetFile =>
 }
 
 val sendReport = SendReport { newReportFile =>
-  sendReport(newReportFile)
+  sendReportToDashboard(newReportFile)
```
Comment (jonas) [inline README.md:31]: @dmitri-carpov Renamed the method called by this stage and added the stage to the execution plan below

**Comment by yawaramin** [comment]
+1 conditional on my above comments being resolved.

File: doc/process-stages.md (hunk 6 above)
Comment (jonas) [inline doc/process-stages.md:40]: Even if this uses the `LocalDateTime` API I think it is good enough for pseudo-code.

File: doc/time-management.md (hunk 7 above)
Comment (jonas) [inline doc/time-management.md:14]: Added a task to the "Output" section of https://github.com/ypg-data/eigenflow/issues/7

File: doc/process-stages.md (hunk 6 above)
Comment (dmitri-carpov) [inline doc/process-stages.md:40]: agree, I'd keep it as a pseudo-code which eventually will turn in a real one (when #7 is resolved)

File: doc/getting-started.md (hunk 5 above)
Comment (jonas) [inline doc/getting-started.md]: Done in 4cc037b

File: doc/error-handling.md (hunk 4 above)
Comment (jonas) [inline doc/error-handling.md]: Done in 4cc037b

File: doc/error-handling.md (hunk 3 above)
Comment (jonas) [inline doc/error-handling.md]: Done in 4cc037b

File: devops/README.md (hunk 2 above)
Comment (jonas) [inline devops/README.md]: Done in 4cc037b

File: devops/README.md (hunk 1 above)
Comment (jonas) [inline devops/README.md]: Done in 4cc037b

**Pull request merged**